src/hidden.rs

```diff
@@ -20,74 +20,248 @@
+            fn from(t: $type) -> Self {
+                Self {
+                    data: Hidden::hide(t),
+                }
+            }
+        }
+    };
+}
+
+/// A generic type for data that needs to be kept hidden and zeroized when out of scope, and is accessible only by
+/// reference.
+///
+/// You can define a hidden type using any underlying data type that implements `Zeroize`.
+/// This is the case for most basic types that you probably care about.
+///
+/// Hidden data has useful properties:
+/// - The data is not subject to `Debug` or `Display` output, which are masked.
+/// - The data can only be accessed by (immutable or mutable) reference, and cannot be copied.
+/// - The data zeroizes when dropped, and can also be manually zeroized.
+/// - Cloning is safe.
+///
+/// Hidden data supports transparent deserialization, but you'll need to implement serialization yourself if you need
+/// it.
+///
+/// ```edition2018
+/// # use tari_utilities::hidden::Hidden;
+/// # use zeroize::Zeroize;
+///
+/// // In this example, we need to handle secret data of type `[u8; 32]`.
+///
+/// // We can create hidden data from existing data; in this case, it's the caller's responsibility to make sure the existing data is handled securely
+/// let hidden_from_data = Hidden::<[u8; 32]>::hide([1u8; 32]);
+///
+/// // We can access the hidden data as a reference, but not take ownership of it
+/// assert_eq!(hidden_from_data.reveal(), &[1u8; 32]);
+///
+/// // We can create default hidden data and then modify it as a mutable reference; this is common for functions that act on data in place
+/// let mut hidden_in_place = Hidden::<[u8; 32]>::hide([0u8; 32]);
+/// let hidden_in_place_mut_ref = hidden_in_place.reveal_mut();
+/// *hidden_in_place_mut_ref = [42u8; 32];
+/// assert_eq!(hidden_in_place.reveal(), &[42u8; 32]);
+///
+/// // Cloning is safe to do
+/// let mut clone = hidden_in_place.clone();
+/// assert_eq!(hidden_in_place.reveal(), clone.reveal());
+///
+/// // You can manually zeroize the data if you need to
+/// clone.zeroize();
+/// assert_eq!(clone.reveal(), &[0u8; 32]);
+/// ```
+#[derive(Clone, Deserialize)]
 #[serde(transparent)]
-pub struct Hidden<T> {
-    inner: T,
+pub struct Hidden<T>
+where T: Zeroize
+{
+    inner: Box<T>,
 }
 
-impl<T> Hidden<T> {
-    /// Hides a value.
+impl<T> Hidden<T>
+where T: Zeroize
+{
+    /// Create new hidden data from the underlying type
     pub fn hide(inner: T) -> Self {
-        Self { inner }
-    }
-
-    /// Returns ownership of the inner value discarding the wrapper.
-    pub fn into_inner(self) -> T {
-        self.inner
+        Self { inner: Box::new(inner) }
     }
 
-    /// Provides access to the inner value (immutable).
+    /// Reveal the hidden data as an immutable reference
     pub fn reveal(&self) -> &T {
-        &self.inner
+        self.inner.deref()
     }
 
-    /// Provides access to the inner value (mutable).
+    /// Reveal the hidden data as a mutable reference
```

Comment:
Would be good to have a description on why this is safe, given the way `Box` is treated by the compiler when dereferenced. Same for `reveal`.

Reply:
From the analysis @CjS77 did on the compiled version, it's unclear precisely if/where this can go wrong. My understanding was that asterisk dereferencing of a `Box<T>` can perform a move, but need not do so in all cases. One apparent way to avoid this may be to replace any asterisk dereferencing with explicit calls to `<Box<T> as Deref>::deref` instead, but I have not tested this.